cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMetadataTable.java

```diff
@@ -77,35 +74,62 @@ public TopicMetadata getMetadata(TopicId topicId) throws IOException, TopicNotFo
-    byte[] rowKey = MessagingUtils.toRowKeyPrefix(topicMetadata.getTopicId());
-    Put put = tableUtil.buildPut(rowKey)
-      .add(columnFamily, COL, Bytes.toBytes(GSON.toJson(new TreeMap<>(topicMetadata.getProperties()), MAP_TYPE)))
+    byte[] rowKey = MessagingUtils.toMetadataRowKey(topicMetadata.getTopicId());
+    PutBuilder putBuilder = tableUtil.buildPut(rowKey);
+
+    Get get = tableUtil.buildGet(rowKey)
+      .addFamily(columnFamily)
       .build();
 
-    if (!hTable.checkAndPut(rowKey, columnFamily, COL, null, put)) {
-      throw new TopicAlreadyExistsException(topicMetadata.getTopicId());
+    Result result = hTable.get(get);
+    byte[] value = result.getValue(columnFamily, COL);
+    if (value == null) {
+      putBuilder.add(columnFamily, COL, Bytes.toBytes(GSON.toJson(topicMetadata)));
+      if (!hTable.checkAndPut(rowKey, columnFamily, COL, null, putBuilder.build())) {
+        throw new TopicAlreadyExistsException(topicMetadata.getTopicId());
+      }
+    } else {
+      TopicMetadata metadata = GSON.fromJson(Bytes.toString(value), TopicMetadata.class);
+      if (!metadata.isDeleted()) {
+        throw new TopicAlreadyExistsException(metadata.getTopicId());
+      }
+
+      int newGenerationId = (metadata.getGeneration() * -1) + 1;
+      TopicMetadata newMetadata = new TopicMetadata(metadata.getTopicId(), metadata.getProperties(), newGenerationId);
+      byte[] newValue = Bytes.toBytes(GSON.toJson(newMetadata));
+      putBuilder.add(columnFamily, COL, newValue);
+      if (!hTable.checkAndPut(rowKey, columnFamily, COL, value, putBuilder.build())) {
+        throw new TopicAlreadyExistsException(topicMetadata.getTopicId());
+      }
     }
   }
 
   @Override
   public void updateTopic(TopicMetadata topicMetadata) throws TopicNotFoundException, IOException {
     boolean completed = false;
 
-    // Keep trying to update
-    byte[] rowKey = MessagingUtils.toRowKeyPrefix(topicMetadata.getTopicId());
+    byte[] rowKey = MessagingUtils.toMetadataRowKey(topicMetadata.getTopicId());
+    TopicMetadata oldMetadata = getMetadata(topicMetadata.getTopicId());
+    TopicMetadata newMetadata = new TopicMetadata(topicMetadata.getTopicId(), topicMetadata.getProperties(),
+                                                  oldMetadata.getGeneration());
     Put put = tableUtil.buildPut(rowKey)
-      .add(columnFamily, COL, Bytes.toBytes(GSON.toJson(new TreeMap<>(topicMetadata.getProperties()), MAP_TYPE)))
+      .add(columnFamily, COL, Bytes.toBytes(GSON.toJson(newMetadata)))
       .build();
 
+    // Keep trying to update
     while (!completed) {
-      TopicMetadata oldMetadata = getMetadata(topicMetadata.getTopicId());
-      byte[] oldValue = Bytes.toBytes(GSON.toJson(new TreeMap<>(oldMetadata.getProperties()), MAP_TYPE));
+      oldMetadata = getMetadata(topicMetadata.getTopicId());
+      byte[] oldValue = Bytes.toBytes(GSON.toJson(oldMetadata));
```

Comment:
That is already done in the getMetadata call. It will throw TopicNotFoundException if the generation is negative (ie, exists call returns false)